### PR TITLE
3636 Reorganize Tahoe-LAFS manual's main page

### PR DESCRIPTION
**Pull request by sajith**
Ticket is https://tahoe-lafs.org/trac/tahoe-lafs/ticket/3636.  

I humbly submit that the manual page in [this version](https://tahoe-lafs.readthedocs.io/en/tahoe-lafs-1.15.0/) is hard to navigate, and [this version](https://tahoe-lafs--1030.org.readthedocs.build/en/1030/) is easier.

Changes:
- Some top-level categories have been added (running, tech notes, specs, notes for contributors, etc.)
- Rearranged order of documents to fit the categories mentioned above.  (No document should be missing as a result of this reordering, or Sphinx would complain about them missing from the toctree.)
- Depth of the toctree has been limited to 1.  A depth of 2 is harder to navigate.
- Removed "Indices and tables", as it does not do anything presently.  It must have been added by `sphinx-autogen` when generating the new Sphinx doc project.

### CODE REVIEW

File: docs/index.rst
Comment (exarkun) [inline docs/index.rst:8]: I might leave "cloud" out here.  "decentralized" probably gives enough of the right flavor?  Or maybe this text should be exactly the text from the "about" doc (either that text copied here, this text copied there, or some sphinx trick to render the text in both places).

File: docs/index.rst
Comment (exarkun) [inline docs/index.rst:16]: This section feels pretty weak.  I wonder if there's a way to get rid of it or flesh it out somewhat.

File: docs/index.rst
Comment (exarkun) [inline docs/index.rst:22]: How about "Using" instead of "Running"?  If I am interested in the software I know I want to use it.  But in addition to runnin it, it seems I also need to install it, configure it, etc.

File: docs/index.rst
Comment (exarkun) [inline docs/index.rst:37]: Technical notes for whom?  Developers?  End-users?  I think the docs here now are a mix.  For example, if I want to use Tahoe-LAFS maybe I want to configure and run a helper or customize my convergence secret.

The "Running[/Using] Tahoe-LAFS" section above strikes me as a good kind of organization for the documentation.  If I am new to Tahoe-LAFS and approaching the docs, at some point I will decide I want to use it and that section title guides me to exactly the part of the docs I need to look at.

How can we do something similar for this section title (and which docs in it should we move elsewhere to accommodate that)?

File: docs/index.rst
Comment (exarkun) [inline docs/index.rst:63]: Maybe "Contributing to Tahoe-LAFS"?

File: newsfragments/3636.other
Comment (exarkun) [inline newsfragments/3636.other:1]: This could be a `.documentation` fragment now, right?

**Review by exarkun — DISMISSED** [review]
Thanks.

I left some comments inline.  Taking in the bigger picture, it would be nice to have the organizing principles for this index recorded - perhaps in the index itself.  If knowing those principles would help readers navigate the index, great, let's make the information a visible part of the docs.  If it is really only useful to future documentation maintainers, we can still include it as rst comments.

The PR description seems to contain some of the organizing principles, so including those would be a start.  It would be nice if every content change being made could be traced directly to a documented organizing principle, though, so the PR description may just be a start.

Thanks again.

File: docs/index.rst
Comment (sajith) [reply docs/index.rst:16]: How about just "Introduction"?

File: docs/index.rst
Comment (sajith) [reply docs/index.rst:37]: Yes, it is a weird mix.  I believe documents under this section would be of interest to both developers and some end-users (curious minds, people who want to integrate Tahoe-LAFS with some other system, system administrators who needs to know Tahoe-LAFS better, potential contributors, etc.)

Can we call this section "Tahoe-LAFS in depth"?

Some of the things here ("Avoiding Write Collisions", "Things to be Careful..." etc) could be moved elsewhere, and/or combined, but that probably should be separate PR.

File: docs/index.rst
Comment (sajith) [reply docs/index.rst]: Dropped "cloud" in 14fa9ed.  I copied that text from the website.

I don't know if Sphinx can include a fragment from another doc, but it has something called "substitutions".  I am not too sure about using that here.

https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#substitutions

File: docs/index.rst
Comment (sajith) [reply docs/index.rst]: Did this in cf9b388.

File: docs/index.rst
Comment (sajith) [reply docs/index.rst]: Did this in f9e3de1.

File: newsfragments/3636.other
Comment (sajith) [reply newsfragments/3636.other]: Renamed it in 3d9ccde.  The PR that added `.documentation` was added in a PR that started after this.

File: docs/index.rst
Comment (sajith) [reply docs/index.rst]: Sphinx can include portions of another document, like this:

```
.. include:: about.rst
   :start-line: 9
   :end-line: 20
```

That will include the below text from `about.rst` in index page:

```
Welcome to Tahoe-LAFS_, the first decentralized storage system with
*provider-independent security*.

Tahoe-LAFS is a system that helps you to store files. You run a client
program on your computer, which talks to one or more storage servers on other
computers. When you tell your client to store a file, it will encrypt that
file, encode it into multiple pieces, then spread those pieces out among
multiple servers. The pieces are all encrypted and protected against
modifications. Later, when you ask your client to retrieve the file, it will
find the necessary pieces, make sure they haven't been corrupted, reassemble
them, and decrypt the result.

```

That is a little too wordy.  The text that is already here (borrowed from [README.rst](https://github.com/tahoe-lafs/tahoe-lafs/blob/master/README.rst)) feels shorter, and maybe that would be the right text to have on this page:

```
Tahoe-LAFS is a Free and Open decentralized storage system.
It distributes your data across multiple servers.
Even if some of the servers fail or are taken over by an attacker,
the entire file store continues to function correctly,
preserving your privacy and security.
```

Or maybe the actual content of this text should be outside the scope of this PR?  We could not add any new text to the index page in this PR, and just stick with reorganization of table of contents.

File: docs/index.rst
Comment (sajith) [reply docs/index.rst]: Renamed the section to "Introduction" in 0f7e37d.

File: docs/index.rst
Comment (YashNRam13) [inline docs/index.rst:17]: I would recommend renaming this to "Getting Started with Tahoe-LAFS" and include installation, and basic configuration articles". The "Using Tahoe-LAFS" section can cater to parallel stuff like some of the stuff you have mentioned (anonymizing network,CLI commands, etc).

File: docs/index.rst
Comment (YashNRam13) [inline docs/index.rst:24]: I've removed this article and another one (OS-X packaging) in my PR. I've also made some changes to Index file (brought all installation guides at the top). I am not sure how we resolve conflicts which will arise.

File: docs/index.rst
Comment (YashNRam13) [reply docs/index.rst:37]: I would rename this to "About Tahoe-LAFS" and if there are many parallel articles with varying complexity, then maybe "About Tahoe-LAFS in depth".

File: docs/index.rst
Comment (YashNRam13) [inline docs/index.rst:70]: Wouldn't desert-island be a part of installation?

File: docs/index.rst
Comment (YashNRam13) [inline docs/index.rst:5]: I really like this organization of the index page. Everything is more searchable and structured.

File: docs/index.rst
Comment (sajith) [reply docs/index.rst]: Renamed it to "Tahoe-LAFS in Depth".

File: docs/index.rst
Comment (sajith) [reply docs/index.rst:13]: I would remove the blurb if there's some reasoning on why it is unnecessary. :-)

As to why it is probably necessary: I'd imagine that a casual reader (such as someone visiting https://tahoe-lafs.readthedocs.io for the first time) would want to know, in a glance, just what it is that they are looking at.

File: docs/index.rst
Comment (sajith) [reply docs/index.rst:17]: I'm not sure about this.  Introduction is, well, an introduction, and the rest (installing, building, configuring, running, other usage notes) should be under "Using Tahoe-LAFS".

Please take a look at [this build of the docs](https://tahoe-lafs--1030.org.readthedocs.build/en/1030/).  Would you still re-arrange it that way?

File: docs/index.rst
Comment (sajith) [reply docs/index.rst]: I have updated the PR.

File: docs/index.rst
Comment (sajith) [reply docs/index.rst]: That is actually geared towards Tahoe-LAFS hackers, according to the document itself:

> Here’s the story: you leave for the airport in an hour, you know you want to do some Tahoe hacking on the flight. What can you grab right now that will let you install the necessary dependencies later, when you are offline?

I don't think it should be part of installation.  In the current commit (d91638d) after merging master, it is under installation, but I'll move it back under developer notes if you agree about that choice.

File: docs/index.rst
Comment (sajith) [reply docs/index.rst:5]: Thanks. :-)

File: docs/index.rst
Comment (YashNRam13) [reply docs/index.rst:13]: I think the Index page can just give people insight into where they can find information. If it's a casual/first-time reader, they most certainly will visit the "Welcome to Tahoe-LAFS" page. And they are bound to read the same/similar intro paragraphs, which IMO, is undesirable.

File: docs/index.rst
Comment (YashNRam13) [reply docs/index.rst:17]: Yes, thanks for the new build. I would definitely re-arrange it that way. I am not a big fan of having an Introduction section. Feels very retro to me, like we're documenting for print media. A lot of cloud tech companies are still bound to this ideology. See "[The grand theory of Documentation](https://documentation.divio.com/)". I am not a big fan of this and believe this does not work with writing for the internet. Instead, we can cut the clutter and just have two types of articles "About Tech1/Feature1". How to do "Tech1/Feature1".

Sorry about the tangent, but that is the reason I don't want to call it Introduction is because for people to read introductions they have to be motivated/excited about what they're reading about. 95% of the time the user is at help because they couldn't make stuff work or just want to learn **how to do specific things**. And that's the reason I want all titles to (somewhat) be verbs. i.e.,

- About Tahoe-LAFS > Welcome to Tahoe-LAFS
- Getting Started > Introduction
- Installing Tahoe-LAFS > Installation Guide
- Running Tahoe-LAFS > How to Run Tahoe-LAFS

Of course, there are going to be some theoretical articles, but that's why we will have "About ..." articles; where we will actively drive the user to a "How-to" article.

About Nodes will (potentially) redirect users to:
- Configuring a Tahoe-LAFS node
- Running Tahoe-LAFS

I hope that makes sense.

File: docs/index.rst
Comment (YashNRam13) [reply docs/index.rst]: Awesome!

File: docs/index.rst
Comment (YashNRam13) [reply docs/index.rst]: I think we can keep it near installation. We need to rework it along the lines of "Installing Tahoe-LAFS completely offline". But until we get to revamping that, we can keep it under developer notes.

**Review by YashNRam13 — APPROVED** [review]

**Comment by May-Lee** [comment]
@exarkun were your requested changes addressed? @sajith @YashNRam13 does anyone object if I optimistically merge this PR?

**Comment by exarkun** [comment]
> @exarkun were your requested changes addressed? @sajith @YashNRam13 does anyone object if I optimistically merge this PR?

I would like to see my top-level review comment addressed:

> Taking in the bigger picture, it would be nice to have the organizing principles for this index recorded - perhaps in the index itself. If knowing those principles would help readers navigate the index, great, let's make the information a visible part of the docs. If it is really only useful to future documentation maintainers, we can still include it as rst comments.
> The PR description seems to contain some of the organizing principles, so including those would be a start. It would be nice if every content change being made could be traced directly to a documented organizing principle, though, so the PR description may just be a start.

It doesn't matter to me very much *what* those principles are but I don't see how we can collaborate on the documentation without a shared understanding of them.

**Comment by YashNRam13** [comment]
>  Taking in the bigger picture, it would be nice to have the organizing principles for this index recorded - perhaps in the index itself. If knowing those principles would help readers navigate the index, great, let's make the information a visible part of the docs. If it is really only useful to future documentation maintainers, we can still include it as rst comments.

@exarkun I agree with you; the documenting/organization principles for this index need to be recorded. However, I do not think that it needs to be realized by adding comments in the index itself. I'll be building off @sajith 's beautiful index here in this PR as I make changes to the documentation. We can then later work on a style guide which accommodates for this index's organization principles.

>  The PR description seems to contain some of the organizing principles, so including those would be a start. It would be nice if every content change being made could be traced directly to a documented organizing principle, though, so the PR description may just be a start.

The PR description does have some of the principles, but they are highly subject to change as I find out more and more about Sphinx's and Tahoe's capabilities. As a stop-gap solution until we come up with our very own documentation style-guide; we can follow the [WriteTheDocs Style Guidelines](https://www.writethedocs.org/guide/writing/style-guides/). This is not tailored for us but accommodates for everything.

I've created a [new trac ticket](https://tahoe-lafs.org/trac/tahoe-lafs/ticket/3796) for the style guide. @exarkun if this is satisfactory, I'd love for this PR to be merged so that I can use the new index as a base for modifications.

File: docs/index.rst
Comment (sajith) [reply docs/index.rst:20]: @YashNRam13 Made a change in f39d765.  The other changes you describe will have to be separate pull requests, right?

File: docs/index.rst
Comment (blaisep) [reply docs/index.rst:5]: I agree!

File: docs/index.rst
Comment (blaisep) [reply docs/index.rst:16]: For the moment, IMHO, it's best to assume that the reader may not know anything at all about the project (or may be arriving from a page that is too detailed to give a big picture.

File: docs/index.rst
Comment (blaisep) [reply docs/index.rst:20]: I'm 100% in agreement with the structure:
> About Tahoe-LAFS > Welcome to Tahoe-LAFS
> Getting Started > Introduction
> Installing Tahoe-LAFS > Installation Guide
> Running Tahoe-LAFS > How to Run Tahoe-LAFS

**Review by blaisep — APPROVED** [review]
IMHO, we should merge this because it's a big step in the direction we want and implements many of the changes we have been discussing. Thank you @sajith

**Comment by sajith** [comment]
@blaisep Thank you for helping to pick this up again!  I'm away this week.  I'll try to update this PR when I'm back.

What is a few weeks for a three year old PR anyway? ;-)

**Comment by hacklschorsch** [comment]
@sajith Hey!  Good to see you around!  Right, no hurries.

I went ahead and merged in current master in [hacklschorsch:3636.doc-toc-reorg-update](https://github.com/hacklschorsch/tahoe-lafs/tree/3636.doc-toc-reorg-update) / blaisep#4.

@blaisep I'll give including the principles a try!

**Comment by hacklschorsch** [comment]
> I agree that the contributor guide should include a "style guide" section. @hacklschorsch , since @sajith might not be available, can you implement the change @exarkun suggested?

@sajith [already implemented](https://github.com/tahoe-lafs/tahoe-lafs/pull/1030#issuecomment-947008447) (very well IMHO!) [@exarkun's Aug 13, 2021 suggestion ](https://github.com/tahoe-lafs/tahoe-lafs/pull/1030#issuecomment-898519764) in aa391be5f8e04f4cb53040c4451858f7de1e6bc0 and 2b2c1e0eb7cb53f24d4fd7bcfdfee419a173ef0a on Oct 19, 2021.

File: docs/README.txt
Comment (hacklschorsch) [inline docs/README.txt:74]: I like this a lot!

File: docs/index.rst
Comment (hacklschorsch) [reply docs/index.rst:20]: I do like the new structure!

File: docs/index.rst
Comment (hacklschorsch) [inline docs/index.rst:26]: `install-tahoe` seems to be in conflict with the other three `install-on-*` documents. I find it confusing to offer two conflicting ways of installing something.

But really the conflict is in naming only: These three are called "install" but to me look more like "build" documentation.  

May be better suited with the developer docs? I'd move them like in fad99192080e0d578d70c65c676aab27234b8c35.

**Review by hacklschorsch — APPROVED** [review]

File: docs/index.rst
Comment (hacklschorsch) [reply docs/index.rst]: Keeping the 'build' with the 'install' documentation offers the user two incompatible ways of installation.
This choice takes from the documentation's clarity. Let's not make readers reason about things they probably do not know about yet (which method is the one for me?  when should I switch from one guide to the other?  do I have to undo all my steps if it turns out to not be working for some reason?  can i switch between both ways in the middle?).

I would put the 'build from source' with the developer documentation, even if building is a legit way of installation. Please see [my comment and commit below](https://github.com/tahoe-lafs/tahoe-lafs/pull/1030#discussion_r1716088809).

**Review by meejah — APPROVED** [review]
LGTM

**Comment by meejah** [comment]
I resolved the "index" conflicts with the web-editor, so if anyone has a different opinion on their resolution please go ahead.

I intend to merge this "soon", I think any of the vague suggestions / future improvements can be better addressed as new PRs on top. If anyone is motivated to work on the "principals" some more that should be done first please.

**Comment by meejah** [comment]
(Oh, fun, it seems I didn't do enough clicking to satisfy the web editor .... hrmm)

**Comment by sajith** [comment]
Neat, I can continue procrastinating. Thank you, @meejah! :-)

What is the deal with CircleCI?

**Comment by meejah** [comment]
@sajith see https://tahoe-lafs.org/trac/tahoe-lafs/ticket/4098 but basically "it's broken, ignore it for now"

**Comment by hacklschorsch** [comment]
This is online now at https://tahoe-lafs.readthedocs.io/en/latest/ !  This makes me happy. 🎉

**Pull request merged**
